tests/xtro-sharpie/DeprecatedCheck.cs

```diff
@@ -0,0 +1,164 @@
+
+				// If the entire type is deprecated, call it good enough
+				if (HasAnyDeprecationAttribute (managedType.CustomAttributes))
+					return;
+
+				var matchingMethod = managedType.Methods.FirstOrDefault (x => x.GetSelector () == selector && x.IsPublic);
+				if (matchingMethod != null)
+					ProcessItem (matchingMethod, fullname, objcVersion, framework);
+			}
+		}
+
+		public void ProcessItem (ICustomAttributeProvider item, string itemName, VersionTuple objcVersion, string framework)
+		{
+			// Our bindings do not need have [Deprecated] for ancicent versions we don't support anymore
+			if (VersionHelpers.VersionTooOldToCare (objcVersion))
+				return;
+
+			// In some cases we've used [Advice] when entire types are deprecated
+			if (HasAnyAdviceAttribute (item))
+				return;
+
+			if (!HasAnyDeprecationAttribute (item.CustomAttributes))
+			{
+				Log.On (framework).Add ($"!deprecated-attribute-missing! {itemName} missing a [Deprecated] attribute");
+				return;
+			}
+
+			// Don't version check us when Apple does __attribute__((availability(macos, introduced=10.0, deprecated=100000)));
+			if (objcVersion.Major == 100000)
+				return;
+
+			// Some APIs have both a [Deprecated] and [Obsoleted]. Bias towards [Obsoleted].
+			Version managedVersion;
+			bool foundObsoleted = AttributeHelpers.FindObsolete (item.CustomAttributes, out managedVersion);
+			if (foundObsoleted) {
+				if (managedVersion != null && !ManagedBeforeOrEqualToObjcVersion (objcVersion, managedVersion))
+					Log.On (framework).Add ($"!deprecated-attribute-wrong! {itemName} has {managedVersion} not {objcVersion} on [Obsoleted] attribute");
+				return;
+			}
+
+			bool foundDeprecated = AttributeHelpers.FindDeprecated (item.CustomAttributes, out managedVersion);
+			if (foundDeprecated && managedVersion != null && !ManagedBeforeOrEqualToObjcVersion (objcVersion, managedVersion))
+				Log.On (framework).Add ($"!deprecated-attribute-wrong! {itemName} has {managedVersion} not {objcVersion} on [Deprecated] attribute");
+		}
+
+		public static bool ManagedBeforeOrEqualToObjcVersion (VersionTuple objcVersionTuple, Version managedVersion)
+		{
+			// Often header files will soft deprecate APIs versions before a formal deprecation (10.7 soft vs 10.10 formal). Accept older deprecation values
+			return managedVersion <= VersionHelpers.Convert (objcVersionTuple);
+		}
+
+		static bool HasAnyAdviceAttribute (ICustomAttributeProvider item)
+		{
+			if (AttributeHelpers.HasAdviced (item.CustomAttributes))
+				return true;
+
+			// Properites are a special case for [Advice], as it is generated on the property itself and not the individual get_ \ set_ methods
+			// Cecil does not have a link between the MethodDefinition we have and the hosting PropertyDefinition, so we have to dig to find the match
+			if (item is MethodDefinition method) {
+				PropertyDefinition property = method.DeclaringType.Properties.FirstOrDefault (p => p.GetMethod == method || p.SetMethod == method);
```

Comment:
for performance reasons [1] don't access Cecil collections without checking the `Has*` property, e.g.
```
if (method.DeclaringType.HashProperties) {
...
}
```

[1] if empty it will allocate an empty collection (and there's ton of metadata, so it can slows down execution quite a bit)